torch/distributed/_tensor/ops/tensor_ops.py

```diff
@@ -59,6 +57,45 @@ def default_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
     )
 
 
+@register_op_strategy(
+    [
+        aten.equal.default,
+        aten.is_same_size.default,
+    ]
+)
+def equal_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
+    # equal_strategy deals with ops that comparing two tensor, we need to make sure
```

Comment:
is the idea here that before you can perform an equality test on 2 ops, you need 
 - (a) they have the same logical shape/size- otherwise can skip comparing data
 - (b) if a dim is sharded for one arg, but replicated for another arg, its nots straghtforward to memcmp their content locally
 - (c) so therefore we canonicalize the args by making them all replicated, so they are easy to compare locally


if so, then i wonder if its more optimal to consider for (b) if both input args are sharded, but already have same sharding, could you skip the replicate step and do the comparison on a local-shard basis?

Reply:
Yep this is a great summary, I'll put these into comments too!

> (a) they have the same logical shape/size- otherwise can skip comparing data
Yep this is a good point we can do such a optimization similar to is_same_size! but only compare the data when logical shape are the same

> i wonder if its more optimal to consider for (b) if both input args are sharded, but already have same sharding, could you skip the replicate step and do the comparison on a local-shard basis?

Yes this is what we are doing currently, currently I'm performing this check inside [sharding propagation](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/sharding_prop.py#L200), if all inputs are sharded in the same way, we don't need to do resharding to turn them as replicated. This strategy mainly try to follow one argument's sharding with the max number of shards, and see if every arg have the same sharding layout, if so, directly compare, otherwise, redistribute to the argument's sharding first.